Implement cucumber tests

ddelnano opened the pull request:
## Todo
### Features
- [x] tcp_server.feature
- [x] execution_order.feature
  - [x] Fix validation steps, all others are passing.
- [x] failing_transaction.feature
- [x] hook_handlers.feature
  - [x] Fix validation steps, all others are passing.
  - ~~Need to implement rpc style communication between server and runner~~ Implemented
- [x] multiple_hookfiles.feature
- [x] Cleanup hooks servers on dredd-hooks-go exit.
### Other things
- [x] Add build to travis

ddelnano committed:
Restructure project

ddelnano committed:
Cucumber tcp_server feature fully passing

ddelnano commented:
@snikch work in progress right now but just opening so the people at apiary can see how close I am.  I am thinking this should be done by the end of the weekend.  Anything that looks bad just ignore until I clean it up after everything is passing :).

snikch commented:
👌

honzajavorek commented:
Wonderful! cc @netmilk @w-vi @abtris

Two pointers:
- We found out there is (at least) one test in the Cucumber test suite, which will never work correctly for Go. From what I understood, Dredd sends dummy object to hooks and expects them to return it back. This works for languages like JS / Python / Ruby, but Go uses types and expects some sort of "transaction" object, not dummy object, so it's not able to just send it back. We should fix the test in the test suite and all its clones so it sends dummy object which at least resembles "transaction" first so it can work for Go. If there's PR on the [hooks template repo](https://github.com/apiaryio/dredd-hooks-template) with fix, I can help with updating all other hooks repos where it's copied.
- I tried to contribute Travis setup, so it should be somewhat worky [here](https://github.com/snikch/goodman/blob/master/.travis.yml).

ddelnano commented:
@honzajavorek I modified the cucubmer test suite to send an empty object and array to get the test passing for the exact problem you mentioned.

Since I am newer to Go I do not know what the best way around this is.  I could write a custom marshaler to parse either this dummy object or a valid transaction object.  I would think its probably unnecessary work and it would be better to update the hooks template repo with a valid object, but whatever we want to do is fine with me.

See here for the change I [made](https://github.com/snikch/goodman/pull/5/files#diff-60124dd45a23a8b2ea2358d2c420d09aR14).

w-vi commented:
What dummy object are you guys talking about? I can't find it. Is it the `data` key?

ddelnano commented:
@w-vi Look at the [difference](https://github.com/apiaryio/dredd-hooks-template/blob/master/features/tcp_server.feature#L14)

ddelnano commented:
Go basically provides a nice way to unseralize json into structs and if that structure is not the same each time it makes the json parsing more difficult so I made it blank so that go would serialize that into nothing since the dummy object does not fit the schema of the transaction object's json.

w-vi commented:
I see, well you don't care about what is in data so just unmarshall it as JSON string, whatever there is the handler doesn't care.

ddelnano commented:
Yea but why I modified it is that perfectly functional code fails the test because I would have to do extra work to make that tcp_server feature pass as its currently written, which is why I modified it for the time being unless there is some compelling reason to make the go hooks pass for some arbitrary "data" json key.

w-vi commented:
Looking at the https://dredd.readthedocs.io/en/latest/hooks/#transaction-object-structure transaction object structure I don't see any `data` key so I guess we can leave it as is.

ddelnano commented:
The transaction object should not have a data key, the data key has a value that **should** be the transaction object.  Thats why it would normally fail.  Instead Go sees an empty object for `"data":{}` so it just turns it into nil instead of blowing up for the bogus value that currently exists in the hooks template repo.

ddelnano commented:
So I ran into a problem during implementing this.

@snikch I noticed that you put the server in the same file that the hooks are defined.  I tried to avoid doing that because that would prevent multiple hooks files from being used.  So I used the rpc package as a workaround.  What I did was the main server file that gets executed runs 2 servers: an rpc server that adds callbacks to the runner and the http server that dredd interacts with.  The hook files then connect to the rpc server and append the callbacks to the runner.  This all worked great and I was able to have multiple hook files, however, when running the test suite I noticed the runner was receiving the callbacks but they were all nil.  This is because the way rpc encodes / decodes data it does not allow function or channel types which means this approach will not work.

It could be implemented the way you started off, but I don't see a way that multiple hook files could be supported.  Do you have any ideas?

@honzajavorek

snikch commented:
Are the hook files complementary, i.e. could you merge them somehow? Alternatively you could move the file to separate folders and then build a binary for each one. That'll start to ramp the time taken to build up though :\

ddelnano commented:
@snikch I don't think it would be possible to merge them.  The way the php hooks work is that the hooks files are `require`d which is basically just like copy and pasting code but since go is compiled I do not believe there is any equivalent to that.

What I described in my previous comment was that the http server and each hook file was its own binary.  The http server also had an additional rpc server that each hook binary was able to connect to.

Actually I think this would work.  Instead of the http server having the rpc server, the http server could act as a rpc client and connect to a rpc server that runs on the hooks binary.  This prevents the previous problem since instead of the hooks trying to serialize the callbacks to the server  (which is not allowed) the server just needs to serialize the dredd transaction. The problem with this approach is that for each dredd transaction the http server would have to make an rpc call to each of the hooks servers (one for each hook file).  There would need to be some kind of service discovery so that the http server would know what port to connect to for all the different hooks servers.  It could be something as simple as start at 61322 (1 after the dredd http port) and just count up.  It feels weird to me that to add another hooks file that a whole new server would be necessary but idk of another way of getting this to work.

ddelnano committed:
Working implementation with rpc server on http server will not work, rebase this later

ddelnano committed:
Update .gitignore

ddelnano commented:
I have most of it implemented for what I described in my last comment.  Hopefully should have these tests passing by tomorrow at some point.

ddelnano committed:
Add partial implementation still WIP

ddelnano committed:
Get feature/failing_transaction.feature passing

ddelnano commented:
@snikch finally pushed some code.  Still new to way packages should be laid out and I moved it around a couple times but left as is for now.

The failing_transaction test works.  The execution_order and hook_handler ones pass if I comment out the assertions on the before each validation and before validation events.  @honzajavorek is there any way that the newest version of dredd is not sending those messages?  I printed out all the messages I am receiving from dredd and I do not see any validation events.

![image](https://cloud.githubusercontent.com/assets/5855593/15527808/ddd0bc8e-220b-11e6-987f-88f38f1cd79f.png)

The multiple hooks has not been implemented yet but should be easy to add.  The part I am struggling with right now is that all the servers that get started up are not cleanly shutdown when dredd closes dredd-hooks-go.  If I run dredd-hooks-go myself and connect it to the hooks server and issue a control C from the terminal everything gets cleaned up properly so I need to figure that out still but it is getting there.

snikch commented on hooks/hooks.go:
```diff
@@ -0,0 +1,241 @@
+	h.afterAll = append(h.afterAll, fn)
+}
+
+// func (h *Hooks) Before(name string, fn Callback) {
```
You can clean up the rcp calls etc for now right? They'll always be in the commit history if you need the code again.

ddelnano replied:
I was gonna rebase this and rewrite the history eventually so its gonna be like a single massive commit but probably better than having this commented stuff in there.

snikch commented on hooks/server_test.go:
```diff
@@ -0,0 +1,171 @@
+var addr = fmt.Sprintf(":%d", port)
+var serv = NewServer(&run, port)
+
+func TestServerRunBeforeAllRPC(t *testing.T) {
```
Remove this test?

ddelnano replied:
Uncommented these since they are useful.

snikch commented on hooks/server_test.go:
```diff
@@ -0,0 +1,171 @@
+	}
+}
+
+func TestServerRunAfterAllRPC(t *testing.T) {
```
And this one?

ddelnano replied:
Same thing for this.

snikch commented on server.go:
```diff
@@ -50,9 +50,11 @@ func (server *Server) Run() error {
 	server.conn = conn
 
 	for {
+		// fmt.Println("Reading from connection")
```
If you don't need these printlns, can you please remove them (below too).

snikch commented on server.go:
```diff
@@ -127,6 +131,40 @@ func (server *Server) ProcessMessage(m *message) error {
 	default:
 		return server.sendResponse(m, m.transaction)
 	}
+	// m.transaction = &t.Transaction{}
```
Remove this unused code too.

snikch commented:
Just had a quick squiz - the code all looks legit, and the package names are good 👌. Just a little bit of clean up comments at this point.

honzajavorek commented:
@ddelnano Interesting. Before releasing every Dredd since the big refactoring in January (even the `pre` versions), we first run it through the cucumber test suite at each hook repo available. That means, I'm pretty confident all versions of Dredd since the beginning of the year are able to pass the same test suite for Py/Rb/PHP/... hooks. Since you're author of the PHP hooks, you can easily compare what happens there and what Dredd normally sends.

I scanned through the screenshot of the output and found no clues on what did/did not happen during the communication :neutral_face: Try to modify the test suite to run dredd with the `silly` log level: https://dredd.readthedocs.io/en/latest/usage-cli/#-level-l - that's what comes to my mind at first place, but I'm not sure it will help much with debugging :hammer:

honzajavorek commented:
And great job by the way! This is really awesome :clap:

ddelnano committed:
Uncomment tests

ddelnano committed:
Remove commented code from server.go

ddelnano commented:
@snikch Yea I apologize for all the commented mess because I had to completely overhaul the implementation details like initially discussed I got so many compilation errors and didn't have anything committed so I didn't want to lose history.

@honzajavorek Like how you suggested about the PHP hooks, I tried to run the cucumber suite with the latest version of dredd since previously I was using the same globally installed version locally (I know travis installs it new each time).  Problem is my gems are all jacked up and when I try to run  my tests locally I get this.

<img width="633" alt="screen shot 2016-05-25 at 7 14 51 am" src="https://cloud.githubusercontent.com/assets/5855593/15538165/75cb9a9a-2248-11e6-955c-5ba326e0ea22.png">

I am not sure why its dieing but when I reinstalled all the gems and I got different output than when I installed them for this project.

<img width="705" alt="screen shot 2016-05-25 at 7 15 05 am" src="https://cloud.githubusercontent.com/assets/5855593/15538180/97c2de56-2248-11e6-9cf6-8f8beb65cdad.png">

Any clues on what is happening there?  I am not very experienced with bundler / installing ruby gems so I am gonna need to look into it more.  I am using the same ruby version (2.3.0 using rvm) and I also noticed my gems were installed in vendor/bundler for the php ones but I think the go ones might be installed in the .rvm directory.

honzajavorek commented:
@ddelnano woohoo, this hits also my Ruby knowledge limitations :open_mouth: I'll have to summon my Apiary Ruby friends to advice on this

ddelnano commented:
@honzajavorek so I finally figured out the ruby problem.  It was due to bundler installing gems in `vendor/bundle` instead of in `.rvm` this was because when I wrote the PHP hooks I did not use rvm and now that I do it was screwed up.  PHP ones pass fine with the latest version of dredd so it seems to be sending the event for the validation.  Running dredd with log level silly didn't seem to provide more useful information so I will have to look into it more.

ddelnano commented:
@honzajavorek So I did some more digging and found out that the tests that fail for the validation hooks the following is happening.

<img width="1270" alt="screen shot 2016-05-26 at 11 49 31 pm" src="https://cloud.githubusercontent.com/assets/5855593/15597263/88957e8e-239c-11e6-8591-d4ec241dc7b8.png">

Dredd seems to be crashing on [this](https://github.com/apiaryio/dredd/blob/master/src/transaction-runner.coffee#L517) line.  Still haven't fully wrapped my head around this code yet but `transaction.request['body']` is coming back undefined and the js that gets compiled from the coffescript does a strict comparison to an empty string.  Not sure if you have any ideas?

honzajavorek commented:
It would be interesting to find out why `transaction.request['body']` evaluates as `undefined`, but the crash seems to be a bug. I think the line should look like this:

``` coffee
req.write(transaction.request.body) if transaction.request.body
```

ddelnano commented:
@honzajavorek I made the comparison less strict ( `!=` ) and it just timed out and never fulfilled.  Didn't really get too much time to look but this weekend I am planning on wrapping up this PR so I am gonna trace back where this is becoming undefined.

I logged the transactions out after they were compiled and the `transaction.request.body` was set to an empty string so somewhere its getting modified.

ddelnano committed:
Cucumber tests passing

ddelnano committed:
Child processes now exit properly

ddelnano committed:
Hooks server accepts its port from -port flag

ddelnano committed:
Update cucumber tests and main file to remove hard coded port

ddelnano committed:
Cleanup comments, TODOs, and other stuff

ddelnano committed:
Binary does not belong in repo

ddelnano committed:
Add travis.yml file

ddelnano committed:
Replace ddelnano with snikch, was just testing file on my travis account

ddelnano committed:
Remove unused variable

ddelnano commented:
@snikch Ok this PR should be considered complete.  There are 2 things I have not completed as of right now, one of which I think can wait.

1.) How does the end user use this and how does dredd find the executable?

Since I am still in the process of learning go best practices I was hoping you would have some ideas?  I was thinking that the end user does `go get github.com/snikch/goodman && go install`.  Dredd would then look for `$GOPATH/bin/goodman` when it executes with the `--language go` flag.

2.) For the multiple hook files I still have not implemented the globing (see [here]()).

I would like to make that work but since I have spent a lot of time on this recently I would like to get this merged as is if possible since this has been going on for a while.  The other change that I think needs to be made with the hookfiles is that it might be better for the wildcard purposes to have the main file compile the hook go files instead of running the binary like they do now. This would allow the wildcard matching to be easier since it can be restricted to files with the go extension rather than trying to look for an executable which would potentially require inspecting the files with executable permissions, executing them and recovering from errors (since they might not actually be the output of a compiled go program).

@honzajavorek I was never able to figure out what was up with dredd but in order to get around the issues with the validation hooks above I had to do [this](https://github.com/snikch/goodman/pull/5/files#diff-160cef1d49980788634ac17a48b25a8bR18).

ddelnano committed:
Update .feature files so that dredd-hooks-go is compiled fresh for each test

ddelnano committed:
Remove unnecessary print statement

ddelnano commented:
Repo with an example using the hooks is [here](https://github.com/ddelnano/dredd-hooks-go-example).  Would most likely want to merge this in to this repo but I had to modify dredd and also instructions on the README would not work until this PR is merged.

honzajavorek commented:
@ddelnano Interesting. I would be curious if we can backport any changes to the test suite back to https://github.com/apiaryio/dredd-hooks-template later on or file some issues there. Maybe in the future someone is going to write another hooks implementation and will bump into similar problems.

ddelnano commented:
@honzajavorek Yea I agree, definitely a good idea.  I will look into that later today.

ddelnano commented:
@honzajavorek there are some additional features I implemented in the php hooks that I was hooking to also backport along with the implementation to the other languages as well.

snikch commented:
Sorry for the absence, just getting back on board after a bit of time off. To answer the question about the executable, you're probably in a better position to answer that than me now as I had envisaged the project as a library rather than a standalone. I'd just expect go `go get github.com/snikch/goodman/bin` it and have it installed, i.e. the binary should be available in the `$PATH` to dredd as `goodman` - does that make sense?

It might be worth following a common pattern and changing `/bin` to `/cmd/goodman` so you can run `go get github.com/goodman/cmd/goodman` and get a binary `goodman` installed in your `$GOPATH/bin` dir.

ddelnano commented:
@snikch well its kind of both (library and executable).  In order to make this work the way all the other hooks are there is the goodman executable which is what dredd talks to but the user also needs to have hookfiles that use the project like a library and the goodman executable uses the hookfiles as RPC servers to execute the transactions.

Is placing the executable at `/cmd/goodman` a go best practice?  Meaning `/cmd/projectname`?  Well if the user did `go get github.com/goodman/cmd/goodman` and get a binary goodman installed in their `$GOPATH/bin` dir like you suggested I am assuming they wouldn't have the rest of the code?  Because the need the whole repo in order to use it like an executable and library like I mentioned above.

[These](https://github.com/ddelnano/dredd/blob/f40079fa74f976672f946f02a8eee143d1cd6a9b/src/hooks-worker-client.coffee#L141-L145) are the instructions I opened up as a PR which worked with the example repo that I posted to above.

snikch commented:
`go get github.com/goodman/cmd/goodman` will get the entire git repo, so that would also make the library available. It would do so even if it was in a different repo too since it's a dependancy.

It's not necessarily `/cmd/projectname` so much as `/cmd/binaryname`, which generally just happens to match the project name. You could also call it `/cmd/goodman-server` or some such.

Rule of thumb is that if the project is primarily a lib, add a `/cmd/` dir with dirs for any binaries. If the project is primarily a command, have it in the root and any library stuff in `/lib` or similar. There's nothing hard and fast but that's often touted as best practice.

### Examples

https://peter.bourgon.org/go-best-practices-2016/#top-tip-2
https://github.com/influxdata/influxdb/tree/master/cmd
https://github.com/codahale/sneaker

ddelnano committed:
Move package command to idiomatic go location

ddelnano committed:
Update tests to use goodman instead of dredd-hooks-go

ddelnano commented:
@snikch ok I updated that like you said.  Let me know if there is anything else I need to do before this can be merged.

snikch commented on cmd/goodman/main.go:
```diff
@@ -0,0 +1,76 @@
+func main() {
+	cmds = make(chan *exec.Cmd, 50)
+	args := os.Args
+	hookPaths := args[1:len(args)]
```
This will panic if the length of args is 0.

ddelnano replied:
Won't it always be 1 since the current binary is always going to be in `os.Args`.

snikch replied:
Correct. Ignore me. I was more thinking around accessing an index after, which `1` would be - but I forgot about the special case for this that Go does. All good.

ddelnano replied:
Always good to be skeptical 👍

snikch commented on cmd/goodman/main.go:
```diff
@@ -0,0 +1,76 @@
+	args := os.Args
+	hookPaths := args[1:len(args)]
+	c = make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
```
You can compact these calls to one line `signal.Notify(c, os.Interrupt, syscall.SIGTERM)`

ddelnano replied:
Nice, I did not look at the function signature that closely when I wrote this.

snikch commented on cmd/goodman/main.go:
```diff
@@ -0,0 +1,76 @@
+	"github.com/snikch/goodman"
+)
+
+var c chan os.Signal
```
FYI you can avoid multiple `var` declarations with parenthesis.

``` go
var (
 foo string
 bar string
)
```

ddelnano replied:
Good call, will update.

ddelnano committed:
Consolidate signal.Notify into a single call

snikch commented on cmd/goodman/main.go:
```diff
@@ -0,0 +1,76 @@
+	for cmd := range cmds {
+		cmd.Process.Kill()
+		count++
+		if hooksServerCount == count {
```
Is this count necessary? Does this code ever get hit except perhaps on the last loop?

From the looks above with the range over `hookPaths` it seems that `hooksServerCount` is always going to be `>= len(cmds)`

ddelnano replied:
@snikch yea what happens is that the program will get stuck trying to read from the channel after all the commands have been killed.  I am sure using a channel to manage this would probably be better but I had a very hard time making sure that when this main server process exits all the hook servers do as well.

The code gets executed if the signals are caught **or** if the execution goes correctly and dredd closes the socket connection.

snikch commented on server.go:
```diff
@@ -73,19 +75,25 @@ func (server *Server) Run() error {
 	}
 }
 
+func (server *Server) CloseRunnerConnections() {
```
Needed?

ddelnano replied:
Nope, removed.

ddelnano committed:
Use more compact var declaration

ddelnano committed:
Remove unused method

snikch commented on server_test.go:
```diff
@@ -0,0 +1,60 @@
+	messages := []struct {
+		Payload []byte
+	}{
+		// TODO: Figure why I could not use `` quoted string
```
`` should have worked fine here. Let me know if you want a hand to figure out what went wrong.

ddelnano replied:
It didn't seem to.  I can try to replace it but I never got the \n characters to come through unless I used `" "`

snikch replied:
Gotcha, it won't interpret special characters - you'd need to literally have a new line.

ddelnano replied:
Ok cool, thanks for the explanation.  I will remove the TODO and keep as is.

snikch commented on test.sh:
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+go build -o bin/dredd-hooks-go
```
Is this still the correct command?

ddelnano replied:
Nope that is not even used anymore, removing.

ddelnano committed:
Remove unused test file

snikch commented on transaction/transaction_test.go:
```diff
@@ -0,0 +1,12 @@
+
+func TestSettingFailToString(t *testing.T) {
+	trans := Transaction{}
+	trans.Fail = "Fail"
```
I'm not sure this test is accomplishing much. Is this more to ensure nobody changes the field in the future (i.e. the test is that the test compiles)?

ddelnano replied:
Yea the reason for it was that since `Fail` was of interface type I was pretty sure it would work ok but wanted to be 100% sure.  I can remove it if you want.  It was mostly me just kind of exploring the interface type.

snikch commented:
All looks pretty solid to me.

ddelnano committed:
Remove TODO, literal new line would have been needed to use backticks

snikch commented on cmd/goodman/main.go:
```diff
@@ -0,0 +1,77 @@
+				}
+			}()
+			// Must sleep so go routine running hooks server has chance to startup
+			time.Sleep(100 * time.Millisecond)
```
Not a blocker, but adding in a sleep is always a code smell to me. No point worrying about it if it's complicated due to the current architecture.

ddelnano replied:
@snikch yea I wasn't too happy with that either.  I would like to make this file more modular and have better tests for it than just the .feature files in the future.

ddelnano commented:
I think I addressed everything except I left in the transaction_test.go file.

ddelnano committed:
Remove TODO file

ddelnano committed:
Update readme to reflect new api

ddelnano committed:
Tweak the README

ddelnano committed:
Fix numbering

ddelnano commented:
@snikch I forgot the README needed to be updated.  Can this be merged now?

snikch commented:
Heh, I forgot I had to do the merging :D
